### PR TITLE
feat(eks): improve security using IRSA conditions

### PR DESCRIPTION
**Pull request by vlesierse**
Currently the `ServiceAccount`construct creates a role with no conditions to the trust relationship or assume role policy. Without this it is possible for other pods in the same namespace to assume the role. To tighten this security the conditions needs to be set.

Documentation: https://docs.aws.amazon.com/eks/latest/userguide/create-service-account-iam-policy-and-role.html#create-service-account-iam-role

- [x] Add condition to the policy document using a custom resource
- [x] Add unit tests
- [x] Add integration tests
- [x] Adjust README and remove warning

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by vlesierse** [comment]
@eladb I've chosen to put the custom resource in the `aws-eks` module as it is currently specific for this use case. Having a patching construct in the `aws-iam` module didn't feel right. I've tried it in this branch: https://github.com/vlesierse/aws-cdk/tree/iam/policytemplate

**Comment by vlesierse** [comment]
Yep, this looks lot better. Probably the lack of knowledge on how the tokenisation works within CDK. I'll dive a bit deep to understand it. Meanwhile I'll do my adjustments when this lands on the `master` branch.

File: packages/@aws-cdk/aws-eks/lib/service-account.ts [hunk 1]
```diff
@@ -61,9 +60,16 @@ export class ServiceAccount extends Construct implements IPrincipal {
     this.serviceAccountName = props.name ?? this.node.uniqueId.toLowerCase();
     this.serviceAccountNamespace = props.namespace ?? 'default';
 
-    this.role = new Role(this, 'Role', {
-      assumedBy: new OpenIdConnectPrincipal(cluster.openIdConnectProvider),
+    const conditions = new CfnJson(this, 'ConditionJson', {
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/service-account.ts:63]: add a reference to aws docs

File: packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/cluster.ts [hunk 2]
```diff
@@ -198,7 +198,7 @@ export class ClusterResourceHandler extends ResourceHandler {
           Endpoint: cluster.endpoint,
           Arn: cluster.arn,
           CertificateAuthorityData: cluster.certificateAuthority?.data,
-          OpenIdConnectIssuerUrl: cluster.identity?.oidc?.issuer,
+          OpenIdConnectIssuerUrl: cluster.identity?.oidc?.issuer?.substring(8), // Strips off https:// from the issuer url
```
Comment (eladb) [inline packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/cluster.ts:201]: This is technically a breaking change. Perhaps the correct approach is to add another attribute `OpenIdConnectIssuer` which will return the value without the `https://` prefix.

Comment (vlesierse) [reply packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/cluster.ts:201]: Ah yeah. Good one. I'll make the change.

Comment (vlesierse) [reply packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/cluster.ts:201]: Should I create the additional property on the Cluster construct as well?

Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster-resource-handler/cluster.ts:201]: Sure

File: packages/@aws-cdk/aws-eks/lib/cluster.ts [hunk 3]
```diff
@@ -708,7 +723,7 @@ export class Cluster extends Resource implements ICluster {
    * @param id the id of this service account
    * @param options service account options
    */
-  public addServiceAccount(id: string, options: ServiceAccountOptions) {
+  public addServiceAccount(id: string, options: ServiceAccountOptions = { }) {
```
Comment (pahud) [inline packages/@aws-cdk/aws-eks/lib/cluster.ts:726]: the `options` is still mandatory here

maybe

```
   public addServiceAccount(id: string, options?: ServiceAccountOptions = { }) { 
```

Comment (eladb) [reply packages/@aws-cdk/aws-eks/lib/cluster.ts:726]: The default value turns it into optional. The “?” Is not needed

**Pull request merged**